### DISCUSSION TITLE
v4.1.0 特性征集与讨论 | v4.1.0 discussion

### DISCUSSION THREAD

**Discussion by Guozhanxin** [RT-Thread/rt-thread#6292] [Ideas]
v4.0.4 已经发布了，所以我们开始筹备下个版本的发布了，下个版本按计划是一个次版本的更新，将会包含更多的新特性。

如果您也有一些功能想法或对当前计划的看法，请发表评论。

v4.0.4 has been released, so we have started to prepare for the release of the next version. The next version is planned to be a minor version update and will contain more new features.

please comment if you also have some features ideas in mind or share your opinion about the planned features.

**Features**

- [ ] MAL (mpu abstract layer) support. https://club.rt-thread.org/ask/article/3188.html
- [x] PSE51 support. https://github.com/RT-Thread/rt-thread/pull/5384
- [x] Update PM Component https://github.com/RT-Thread/rt-thread/pull/5332
- [x] armclang c++11 

**BSP**

- [x] Add STM32L5 https://github.com/RT-Thread/rt-thread/pull/5250
- [x] Add STM32U5 https://github.com/RT-Thread/rt-thread/pull/5280
- [x] Add Renesas CPK-RA6M4 https://github.com/RT-Thread/rt-thread/pull/5262

**Package**

- [x] PikaScript
- [x] TinyUSB

**LVGL** https://github.com/lvgl/lvgl/issues/2790

- [x] STM32F469I-DISC supports for LVGL
- [x] QEMU-A9 supports for LVGL. https://github.com/RT-Thread/rt-thread/pull/5350
- [x] VS2019 simlulator supports for LVGL https://github.com/RT-Thread/rt-thread/pull/5394
- [x] STM32L475-Pandora supports for LVGL https://github.com/RT-Thread/rt-thread/pull/5204
- [x] Nuvoton-N9H30 supports for LVGL https://github.com/RT-Thread/rt-thread/pull/5399
- [x] Nuvoton-NK-980IOT supports for LVGL https://github.com/RT-Thread/rt-thread/pull/5399
- [x] implement documentation

**Exquisite**

- [ ]  ~~Rename finsh to msh~~
- [x] Update the file system elmfat version to R0.14b https://github.com/RT-Thread/rt-thread/pull/5223
- [x] Remove RT_USING_TC https://github.com/RT-Thread/rt-thread/issues/1974
- [x] Organize and clarify the level of DFS, libc and posix

**Others**

- [x] Change the flag of mutex from FIFO to PRIO https://github.com/RT-Thread/rt-thread/pull/5284
- [x] Optimize memory protection mechanism https://github.com/RT-Thread/rt-thread/pull/5175
- [x] NewLib issue
- [x] CFLAG flag issue https://github.com/RT-Thread/rt-thread/pull/4953 https://github.com/RT-Thread/rt-thread/pull/5397

**Comment by enkiller** [comment]
测试框架这部分有啥具体的动作吗？

**Comment by Guozhanxin** [comment]
> 测试框架这部分有啥具体的动作吗？

直接移除掉吧

----

借楼整理一下正式发布前需要解决的问题：

- [x] https://github.com/RT-Thread/rt-thread/issues/5584
- [x] https://github.com/RT-Thread/rt-thread/issues/5323
- [ ] https://github.com/RT-Thread/rt-thread/issues/568
- [x] ~~https://github.com/RT-Thread/rt-thread/issues/5594~~
- [x] https://github.com/RT-Thread/rt-thread/issues/1857
- [x] ~~一个线程获取多个mutex可能存在优先级翻转(v5.0.0)~~
- [x] 完善对cmake等脚本的支持，在4.0.4发布直播时，就有人问cmake的问题，cmake这块应该重视起来，完善一些细节问题，确实有很多人喜欢用cmake https://github.com/RT-Thread/rt-thread/issues/5482
 - [x] CCFLAGS疑似失效 https://github.com/RT-Thread/rt-thread/pull/5694
 - [x] 信号量需要在系统启动前可以使用 https://github.com/RT-Thread/rt-thread/pull/5682
 - [x] 使用SAL，在开启RT_USING_POSIX（或4.1.0之后的RT_USING_POSIX_FILEIO）之后，没有开启SAL_USING_POSIX，会导致两套fd机制共存，造成sal创建socket返回的fd是1，close(1)的时候，走的是DFS那边，把STDIN给关了。理论上这个问题已经在 https://github.com/RT-Thread-packages/nopoll/issues/9 中提及到并修复掉了，但是天翔依然遭遇到了类似的问题。

**Comment by mysterywolf** [comment]
## 需要继续跟踪的PR和论坛帖子
- [ ] #5296 其他系列的STM32是否也有这个问题
- [ ] STM32 LPUART的问题
  - https://club.rt-thread.org/ask/question/433249.html RTT 针对stm32lx系列低功耗串口驱动bug
  - https://club.rt-thread.org/ask/question/424812.html 反馈rtt串口驱动对低功耗串口lpuart1不兼容的问题
  - https://club.rt-thread.org/ask/article/3412.html STM32L4 LPUART1 DMA下不能工作的解决办法
- [x] https://club.rt-thread.org/ask/question/434691.html 论坛反馈的finsh_getchar的一个bug
- [ ] https://club.rt-thread.org/ask/article/54.html STM32 UART
- [ ] https://club.rt-thread.org/ask/question/425502.html STM32L0 drv_usart.c的BUG及解决
- [x] https://club.rt-thread.org/ask/question/429786.html  ulog的问题 https://github.com/RT-Thread/rt-thread/pull/5114
- [ ] https://club.rt-thread.org/ask/question/424912.html drv_pwm.c 有bug
- [x] https://club.rt-thread.org/ask/question/434822.html 同时ulog的timestamp与rtc打印无输出 https://github.com/RT-Thread/rt-thread/pull/5632
- [ ] https://club.rt-thread.org/ask/article/3483.html n32g457 spi驱动
- [ ] https://club.rt-thread.org/ask/article/54.html F030系列芯片UART3~UART6使用问题
- [ ] https://club.rt-thread.org/ask/question/435184.html H7 UART5的问题以及修改方法

## 需要跟踪的其他社区PR或issues
- [ ] https://github.com/apache/incubator-nuttx/pull/5108 Nuttx社区报告

**Comment by liukangcc** [comment]
https://github.com/RT-Thread/packages/tree/24f0a9c9d5d355fd88f416ca8e85c5ba05d52f91/language/pikascript

**Comment by liukangcc** [comment]
CFLAG flag issue 已解决 https://github.com/RT-Thread/rt-thread/pull/5397

**Comment by liukangcc** [comment]
MAL: https://github.com/RT-Thread/rt-thread/pull/5379

**Comment by ImFovery** [comment]
使用rtthread如鲠在喉的一个点：串口终端作为日志输出，即使ulog设置为异步也仍然为阻塞查询输出，占用cpu且输出效率低。

**Comment by evan-liyb** [comment]
希望版本发布后，能与之前的所有包兼容。并且给出从老版本升级到新版本的官方文档。

**Comment by mysterywolf** [comment]
> 前的所有包兼容。并且给出从老版本升级到新版本的官方文档。

https://github.com/RT-Thread/rt-thread/pull/5277
可以看一下这个，新版本有专门的legacy文件夹

**Comment by evan-liyb** [comment]
打开之后，仍然很多不兼容的地方。特别是我自己写的应用层（大概8万行代码），用在新内核上很多地方不兼容了，主要是一些宏之类的做了调整，没去细看为什么要这样改，但希望尽量减少这种调整吧，或者用宏都老的定义管起来，但仍然保留。

**Comment by mysterywolf** [comment]
> 打开之后，仍然很多不兼容的地方。特别是我自己写的应用层（大概8万行代码），用在新内核上很多地方不兼容了，主要是一些宏之类的做了调整，没去细看为什么要这样改，但希望尽量减少这种调整吧，或者用宏都老的定义管起来，但仍然保留。

再不改，里边很多东西都变成毒瘤了。
https://github.com/RT-Thread/rt-thread/issues/5230
https://www.bilibili.com/video/BV15f4y1u7yA

**Comment by thewon86** [comment]
> 使用rtthread如鲠在喉的一个点：串口终端作为日志输出，即使ulog设置为异步也仍然为阻塞查询输出，占用cpu且输出效率低。

请看论坛我发的文章，我刚对串口框架做了个重构，阻塞 io 情况下，降低阻塞时间。

**Comment by thewon86** [comment]
> https://club.rt-thread.org/ask/question/434195.html ps或者list_thread最后一栏错误码，显示的是错误码代码 非常的不友好，别人根本不知道什么意思

所有的 msh 命令，增加 c 函数调用接口，类似 https://gitee.com/thewon/rt_thread_repo/tree/master/user 这里实现的命令一样，既有终端命令入口，也有对等的 c 调用入口。

**Comment by BernardXiong** [comment]
v4.1.0版本release后，可以拉分支出来。然后master也会尝试拉pre-v5.0.0分支，然后把smart相关的一些代码合并。如果春节期间能够合成功，那么再合并回master分支，做为后续的v5.0.x分支而存在。

另外，建议v4.1.0分支也不用做alpha, beta, rc版本，直接出v4.1.0版本，有问题可以快速修正到v4.1.1版本

**Comment by BernardXiong** [comment]
> 打开之后，仍然很多不兼容的地方。特别是我自己写的应用层（大概8万行代码），用在新内核上很多地方不兼容了，主要是一些宏之类的做了调整，没去细看为什么要这样改，但希望尽量减少这种调整吧，或者用宏都老的定义管起来，但仍然保留。

可以一些点上多做沟通，如何在新特性和兼容性上做到平衡

**Comment by mysterywolf** [comment]
> 另外，建议v4.1.0分支也不用做alpha, beta, rc版本，直接出v4.1.0版本，有问题可以快速修正到v4.1.1版本

建议用beta，因为4.1.0 beta版本还有一部分功能原计划4.1.0需要实现的还没有实现。4.0.4已经吃了一次亏了，变化太大，用户反应有些不适应，4.1.0作为大版本，得一次性把较大改变的内容全部都搞定，从4.1.1版本开始就不能剧烈地增加新功能了。否则，如果4.1.0直接发出去了，4.1.1还要增加新功能，可能要重踩4.0.4发布时的问题。

**Comment by mysterywolf** [comment]
- [ ] 加个ARCH_HAS_SMP_FEATURE ，有这个，就可以由用户来开启smp特性，否则没有smp可以供用户选择（熊大提出）
- [x] 用__clang__代替__CLANG_ARM，__CLANG_ARM是我们在rtdef.h里自己定义的宏，使用受限，__clang__是标准的全局宏
- [ ] 移除掉`FINSH_FUNCTION_EXPORT` `FINSH_FUNCTION_EXPORT_ALIAS` 
- [ ] https://club.rt-thread.org/ask/question/434195.html ps或者list_thread最后一栏错误码，显示的是错误码代码 非常的不友好，别人根本不知道什么意思
- [ ] https://github.com/RT-Thread/rt-thread/blob/master/include/rtdef.h#L1063-L1166 需要从rtdef.h里移除掉的 这部分应该属于LCD驱动框架.h负责的 应该是早起驱动框架没成熟的时候塞进去的
- [ ] 筛查github issue中打4.1.0tag的issues https://github.com/RT-Thread/rt-thread/issues?q=is%3Aissue+is%3Aopen+label%3Av4.1.0
- [x] 筛查github pr中打4.1.0tag的PR https://github.com/RT-Thread/rt-thread/pulls?q=is%3Apr+is%3Aopen+label%3Av4.1.0
- [x] _MSC_VE和_WIN32应该什么情况下用哪个 https://github.com/osmcode/libosmium/issues/224#issuecomment-323811903  https://github.com/RT-Thread-packages/lwip/commit/215dec653b3d8bbd514e9ce30392bd807549ebbe
- [ ] https://github.com/RT-Thread/rt-thread/issues/3440#issuecomment-598991797 https://github.com/RT-Thread/rt-thread/issues/4791

**Comment by mysterywolf** [comment]
## Kernel

- [ ] ~~Kernel中有使用int数据类型，应当予以替换掉~~
- [ ] struct thread结构体中采用隐式继承object，应改成显示继承 https://github.com/RT-Thread/rt-thread/pull/5500 (**V5.0.0**)
- [x] struct thread结构体需要清0 https://github.com/RT-Thread/rt-thread/pull/4891
- [ ] timer代码梳理 https://github.com/RT-Thread/rt-thread/pull/5402
- [ ] https://github.com/RT-Thread/rt-thread/issues/1817
- [ ] https://github.com/RT-Thread/rt-thread/blob/master/include/rtdebug.h#L64-L136 这种宏函数的写法会不会造成ROM占用过大?
- [ ] 一个线程获取多个mutex可能存在优先级翻转 (**V5.0.0**)
- [x] https://github.com/RT-Thread/rt-thread/issues/5623 https://github.com/RT-Thread/rt-thread/pull/5637
- [x] 信号量是经过特殊处理的，允许在调度器未初始化之前拿资源为1的情况，因此需要重新处理pr #4740
- [ ] 需要对内核API的调用环境进行检查 https://github.com/RT-Thread/rt-thread/pull/5576 ，抽象成函数，再封装成宏函数，以节约ROM空间

**Comment by mysterywolf** [comment]
## Scons 与 Kconfig
- [x] 完善对cmake等脚本的支持，在4.0.4发布直播时，就有人问cmake的问题，cmake这块应该重视起来，完善一些细节问题，确实有很多人喜欢用cmake https://github.com/RT-Thread/rt-thread/issues/5482
- [x] 补充Sconscript以及scons相关文档，由于RT-Thread采用的Scons作为构建工具本身就很不常见，再加上该构建工具是RTT自己魔改出来的，但是配套的文档并不全面，导致用户无法立即、使用Sconscript
- [x] CCFLAGS疑似失效 https://github.com/RT-Thread/rt-thread/pull/5694
- [x] https://github.com/RT-Thread/rt-thread/issues/5745 对工程的文件排序应该按字母来进行

**Comment by ghost** [comment]
可否提供一个相对统一的 lcd 框架，对上提供统一的接口。这样在 lvgl 包中实现 rtt 的屏幕驱动，底层让用户和 bsp 去对接就好，然后直接编译就可以直接使用了。

**Comment by mysterywolf** [comment]
> 可否提供一个相对统一的 lcd 框架，对上提供统一的接口。这样在 lvgl 包中实现 rtt 的屏幕驱动，底层让用户和 bsp 去对接就好，然后直接编译就可以直接使用了。

这个问题想过，当时rtt自己维护的lvgl软件包（7.0）就是采用直接和rtt lcd框架对接的方式。但是经过实践之后发现，本身lvgl自己就已经提供了底层的抽象层，如果再对接一个lcd框架就显得脱裤子放屁，因此在lvgl官方软件包中，就采用了lvgl原生的api，方便用户做出选择，既可以直接怼lcd原生驱动，也可以自行怼lcd框架。我们这侧就不强制用户非得使用lcd的框架了，给予最大灵活性

**Comment by mysterywolf** [comment]
## exmaple
- [x] 项目实验室软件包 https://github.com/RT-Thread/packages/pull/1189
- [x] utest的头文件路径和group会被误加入到工程中 https://github.com/RT-Thread/rt-thread/pull/5592
- [ ] 将https://github.com/RT-Thread-packages/peripheral-sample
 https://github.com/RT-Thread-packages/filesystem-sample
 https://github.com/RT-Thread-packages/kernel-sample
 https://github.com/RT-Thread-packages/network-sample
 https://github.com/RT-Thread-packages/device-ipc-sample
 请考虑将上述五个仓库的内容并入主仓库example，并重新整理example现存的文件

**Comment by mysterywolf** [comment]
## LWIP
- [x] 移除LWIP 2.0.2
- [x] 优化lwip目录组织结构 https://github.com/RT-Thread/rt-thread/pull/5587 
- [x] 归并整理公共底层代码 https://github.com/RT-Thread/rt-thread/pull/5606
- [x] https://gitee.com/rtthread/rt-thread/pulls/439 LWIP64位兼容  https://github.com/RT-Thread/rt-thread/pull/5606
- [x]  #5045 其他版本的LWIP是否也有同样的问题 https://github.com/RT-Thread/rt-thread/pull/5606
- [x] 推动upstream或与upstream保持同步 https://github.com/RT-Thread-packages/lwip https://github.com/RT-Thread/rt-thread/pull/5629 https://github.com/RT-Thread/packages/pull/1197
- [x] 使用SAL，在开启RT_USING_POSIX（或4.1.0之后的RT_USING_POSIX_FILEIO）之后，没有开启SAL_USING_POSIX，会导致两套fd机制共存，造成sal创建socket返回的fd是1，close(1)的时候，走的是DFS那边，把STDIN给关了。理论上这个问题已经在 https://github.com/RT-Thread-packages/nopoll/issues/9 中提及到并修复掉了，但是天翔依然遭遇到了类似的问题。 https://github.com/RT-Thread/rt-thread/pull/5731
- [ ] https://github.com/RT-Thread/rt-thread/issues/568

------------
- [x] https://github.com/RT-Thread/rt-thread/issues/631
- [x] https://github.com/RT-Thread/rt-thread/issues/1857

**Comment by mysterywolf** [comment]
## LVGL

- [x] https://github.com/RT-Thread/rt-thread/pull/5513#issuecomment-1025073518 pico的屏幕型号并没有指明

**Comment by thewon86** [comment]
https://github.com/RT-Thread/rt-thread/pull/4740#issuecomment-1033219023

**Comment by mysterywolf** [comment]
## 文档/教学视频 (v4.1.0 v5.0.0)
文档是关键！！ 哪怕什么功能都不改进，只补充、丰富文档，看似原地踏步，实则也是坐着直升机原地踏步！
论坛里很多文章应该整合到文档中心，过于分散，论坛中的优秀帖子、文章很容易随着时间的流逝而压箱底，需要将精华整理到文档中心。

![image](https://user-images.githubusercontent.com/34888354/156664982-dca95df8-fb3f-4abc-8e01-77269c4ad984.png)
![image](https://user-images.githubusercontent.com/34888354/156665010-e65296c9-a589-400c-8e7c-742e8196cca4.png)
![image](https://user-images.githubusercontent.com/34888354/156866124-f0786547-f4de-478e-836e-9ff544ca89fd.png)


- [ ] 完善中文文档中心（POSIX、Scons如何使用、如何移植一个BSP等）
- [ ] 英文文档，由海外社区英语母语者重新润色ing https://github.com/RT-Thread/community-activities/blob/master/README.md
- [ ] QEMU和VS模拟器需要给出英文教学文档，方便海外用户在没有开发板的情况下使用RT-Thread
- [ ] https://club.rt-thread.org/ask/question/434766.html
- [ ] 完善LVGL社区文档中RT-Thread移植部分

**Comment by mysterywolf** [comment]
## BSP
- [x] 删除swm320-lq100 https://github.com/RT-Thread/rt-thread/pull/4348#issuecomment-781124528 
- [x] 尝试联系一下这位pr提交作者将pr重新处理一下 https://github.com/RT-Thread/rt-thread/pull/5208
- [ ] 继续推进新版GD32的整合

**Comment by mysterywolf** [comment]
## 命名是个大问题！(v5.0.0)
设备框架文件命名不统一，API命名不统一，很容易把初学者搞蒙 
杨老师那边怎么办？

**无论是宏定义也好、函数名也好、文件也好，一定要有我们自己的标识前缀，因为我们是有众多软件包的，如果名字起的太过于普遍，很容易和其他软件包的名字冲突。**

建议设备框架统一前缀为：dev_pwm.c/.h dev_adc.c/.h

- [ ] kernel 的文件应该有统一的前缀，例如timer mem这些名字很容易和其他文件重名
- [ ] rtdef.h ALIGN->RT_ALIGNED? ALIGN这个名字太common了，会和其他第三方软件包冲突，目前已经和lvgl那边冲了
- [ ] 设备框架 结构体命名不统一
- [ ] 设备框架结构体命名不统一
- [ ] 设备框架函数命名不统一
- [ ] https://github.com/RT-Thread/rt-thread/issues/5287
- [ ] 设备框架文件命名 https://github.com/RT-Thread/rt-thread/issues/3471
- [ ] 由于设备框架命名没有统一前缀（例如dev_xxx），导致adc.c/.h驱动框架和adc.c/.h芯片库冲突（有此风险的还有pin.c adc.c等等）https://club.rt-thread.org/ask/question/434853.html
![image](https://user-images.githubusercontent.com/34888354/156911175-81efab7d-71ff-4b1b-9c1c-32c69a96a25f.png)

**Comment by jifanchn** [comment]
建议考虑CMake作为未来的构建工具，现在越来越多的IDE都在支持CMake了，scons确实有点老了。

目前正在考虑把公司的下一代产品构建在RTThread上，但是这个scons的构建系统让人望而却步。

**Comment by BernardXiong** [comment]
> 建议考虑CMake作为未来的构建工具，现在越来越多的IDE都在支持CMake了，scons确实有点老了。
> 
> 目前正在考虑把公司的下一代产品构建在RTThread上，但是这个scons的构建系统让人望而却步。

可以提一份方案上来，cmake也有些古老，其中的语法方式也不那么友好。scons部分在RTT这边做了挺多的一些调整，有好的有坏的，一下就想更换构建系统，不容易/很不容易。

**Comment by mysterywolf** [comment]
@jifanchn 反对将RT-Thread Scons转移到其他构建方式。这相当于是自废武功。scons构建方式很灵活，基于python语法的优势，很多复杂的文件和路径添加逻辑几句话就可以完成，而且可以引用python丰富的内置函数。Scons是RT-Thread定制化的构建工具，我们可以根据需求自行对scons进行修改，处理更加灵活。
此外RT-Thread支持Keil，IAR VSCODE VS Cmkae Makefile等其他工程的生成，这也全都得益于scons。如果转换到Cmake构建方式，如何构建Keil IAR等等其他的project？Scons是支持生成Cmake工程的，在env里，敲`scons --target=cmake` 即可
目前Scons的主要问题是文档太少了！这也是很多人望而却步的主要原因，实际上常用的Scons函数没有几个，如果有文档或者教学视频，照葫芦画瓢很简单。文档和教学视频是RT-Thread目前依然要急需填的坑！

**Comment by jifanchn** [comment]
@mysterywolf 如果能用SCONS完美构建CMAKE工程也是可以接受的，但是也要考虑到公司的静态库权限管控和SDK升级需求。Keil和IAR这样的IDE我们已经深受其害了，在多人维护的系统和持续迭代更新的产品中，编译目标的控制无法被GIT等源代码版本控制系统维护是造成嵌入式产品源代码碎片化的重要原因之一。另外构建系统不仅需要支持嵌入式代码编译，也要能支持整个软件生命周期的源代码测试用例、并行仿真、第三方算法工程代码生成、开源代码的引用的调用等一系列操作。CMAKE的开发者数量与生态已经不是SCONS能比的了，甚至可以说已经是C/C++事实上的标准构建系统了。想想QMAKE/MAKEFILE，甚至微软VS的构建系统都已经转向支持CMAKE，JB家的CLION甚至只支持CMAKE。换句话说，SCONS生成Keil和IAR这样的工程，也只是对不熟悉开源社区的“传统”嵌入式程序员的兼容罢了。看看还活跃的C/C++开源库，不支持CMake的太少了。缺少说明文档库，其实就是生态缺失的重要标志之一。
RTThread作为国产的优秀、稳定的嵌入式操作系统，不仅应该做到兼容，同时也应考虑能“温和地”正确引领嵌入式开发的发展方向，为嵌入式开发生态转变甚至超越目前的互联网生态模式做出更多努力。
这也是我建议转向CMake的原因之一。

----看看zephyr吧。

**Comment by mysterywolf** [comment]
@jifanchn zephyr的模式学不来的，门槛太高了，得要求使用者有Linux使用经验，全面转到Linux环境或类unix环境下开发，其开发方式很不符合我国国情。

中国开发嵌入式的工程师来自于两部分，一部分是计算机专业以及软件工程专业毕业的，受过Linux等良好的软件开发教育，可以轻松应对 Cmake，makefile，shell等等这些东西。还有一部分是电子、电子、通信专业毕业，其专业内，Linux等软件开发科目并不是必修科目，他们更喜欢用Keil IAR这类集成性IDE来点点点按钮。

当然如果你觉得目前RTT CMake应该如何做，欢迎提出意见，因为我们确实也收到类似希望采用CMake来开发的建议了。但是很难全面转型到Cmake，就像上面说的，我们会完善scons的Cmake脚本生成能力，或者下次发布版本的时候可以把Cmake脚本直接刷到仓库里，让CMake用户可以直接用起来，直接无视Scons，这都是可以讨论的。但是scons肯定是不能动的，我们得保证中国所有专业方向的嵌入式工程师的死活，不能仅仅服务于那一部分非常专业的、熟悉Linux开发环境的嵌入式工程师。

**Comment by jifanchn** [comment]
同意你的说法，keil和IAR是要支持的。

CMake能有某种方法直接和scons并行这是最好的。

互联网最近已经开始卷了，国内未来深入设备底层的开发也会越来越多，我司目前也看到招聘中人员收入，学历和行业的错位。未来也会有越来越多的熟悉Linux的专业程序员进入。

**Comment by mysterywolf** [comment]
# CMake
- 社区已经多次收到希望RTT进一步完善CMake构建的呼声
- 进一步完善通过Scons生成CMake的能力
目前，Scons生成的CMake脚本只是一个文件放在bsp的根目录下，应参考其它CMake构建的开源项目（如zephyr micropython等）的CMakeLists.txt分配模式，可能每个文件夹都有一个CMakeLists.txt，多级，而不是全塞在一个CMakeLists.txt中。
- [x] https://github.com/RT-Thread/rt-thread/issues/5482 https://github.com/RT-Thread/rt-thread/pull/5677
- 考虑这种可能行：上述功能实现的情况下，可以考虑下次版本发布时，直接将CMakeLists.txt刷到各个文件夹下，让熟悉CMake的用户直接用CMake开发，无视掉scons。Scons和Cmake并存，scons负责自动根据sconscript来刷cmakelists.txt

**Comment by aozima** [comment]
- [ ] 支持64位时间戳，以解决千年虫问题
  - [ ] 驱动支持64位时间戳，即使libc还末支持64位
  - [ ] libc支持64位的方案，不同的编译器现在支持的情况不同，可以考虑自己单独维护1套time相关的功能

![2038](https://user-images.githubusercontent.com/1754905/156694274-213ce0ef-2e25-4411-a9c8-73fbf6545bc6.gif)

**Comment by mysterywolf** [comment]
# 软件包
在4.1.0需要升级的：
- [x] https://github.com/RT-Thread-packages/paho-mqtt issue中反馈了一些bug并提出了解决方案，需要整合到仓库中
- [x] https://github.com/RT-Thread-packages/fal FAL根据龙哥建议，应当放到主仓库
- [x] https://club.rt-thread.org/ask/question/433049.html
- [x] https://club.rt-thread.org/ask/question/431576.html 4.1.0 上问题已不存在
- [x] https://club.rt-thread.org/ask/question/430162.html
- [x] https://club.rt-thread.org/ask/question/430162.html
- [ ] libmodbus 转移到RTT官方软件包进行维护 https://github.com/RT-Thread-packages/libmodbus
- [x] https://github.com/RT-Thread/packages/issues/1215

**Comment by BreederBai** [comment]
新版本有计划增加获取USB是否连接的接口吗？USB DTR标志是必须的吗？ChibiOS的usb驱动（STM32F427），上位机勾选DTR，似乎也能通信

**Comment by cha331** [comment]
> # 软件包
> 在4.1.0需要升级的：
> 
> * [ ]  https://github.com/RT-Thread-packages/paho-mqtt issue中反馈了一些bug并提出了解决方案，需要整合到仓库中
> * [ ]  https://github.com/RT-Thread-packages/fal FAL根据龙哥建议，应当放到主仓库
> * [ ]  https://club.rt-thread.org/ask/question/433049.html
> * [ ]  https://club.rt-thread.org/ask/question/431576.html
> * [ ]  https://club.rt-thread.org/ask/question/430162.html

libmodbus软件包编译报错问题也可以修改下，主要涉及几个头文件，论坛里已有解决方案
https://club.rt-thread.org/ask/article/3280.html

**Comment by mysterywolf** [comment]
STM32F1 RTC读写日期不正常 https://club.rt-thread.org/ask/question/434910.html

**Comment by tfx2001** [comment]
> 新版本有计划增加获取USB是否连接的接口吗？USB DTR标志是必须的吗？ChibiOS的usb驱动（STM32F427），上位机勾选DTR，似乎也能通信

@BreederBai 是 RTT 的 USB 协议栈还是 TinyUSB 呢？

**Comment by liukangcc** [comment]
https://github.com/RT-Thread-packages/paho-mqtt/pull/47

**Comment by liukangcc** [comment]
FAL添加到 rt-thread 主仓库：https://github.com/RT-Thread/rt-thread/pull/5662

**Comment by mysterywolf** [comment]
- [x] 在gcc下 uint32_t是long unsigned int ，而在keil下是 unsigned int 。这个不同编译器定义不一样。
rt_uint32_t是 unsigned int，导致rt_uint32_t的函数形参在填入uint32_t的变量时会出现警告。
![30e72baef400ed1b433fdaa38f748f0](https://user-images.githubusercontent.com/34888354/157929024-512eb983-37b6-4d1f-b602-48d6d98145d2.png)

**Comment by mysterywolf** [comment]
## Finsh 疑似bug
- [x] https://club.rt-thread.org/ask/question/434691.html https://github.com/RT-Thread/rt-thread/pull/5676
- [x] ~~https://club.rt-thread.org/ask/question/430704.html~~ 该问题已经不存在

**Comment by mysterywolf** [comment]
![image](https://user-images.githubusercontent.com/34888354/158003128-c3a538cb-272a-415b-ac96-f93670a0ef40.png)
内核可以考虑用这种方式来设置堆栈大小，例如对于龙芯，其idle main线程的栈大小要高于32位的

**Comment by liukangcc** [comment]
> ## Finsh 疑似bug
> * [x]  https://club.rt-thread.org/ask/question/434691.html
> * [x]  https://club.rt-thread.org/ask/question/430704.html

第二个源文件被删了吧 https://club.rt-thread.org/ask/question/430704.html


https://github.com/RT-Thread/rt-thread/pull/5676

**Comment by mysterywolf** [comment]
> 第二个源文件被删了吧 https://club.rt-thread.org/ask/question/430704.html

是的可以不用管了~

**Comment by BreederBai** [comment]
> > 新版本有计划增加获取USB是否连接的接口吗？USB DTR标志是必须的吗？ChibiOS的usb驱动（STM32F427），上位机勾选DTR，似乎也能通信
> 
> @BreederBai 是 RTT 的 USB 协议栈还是 TinyUSB 呢？

@tfx2001  RTT 的 USB 协议栈

**Comment by liukangcc** [comment]
> # 软件包
> 在4.1.0需要升级的：
> 
> * [x]  https://github.com/RT-Thread-packages/paho-mqtt issue中反馈了一些bug并提出了解决方案，需要整合到仓库中
> * [x]  https://github.com/RT-Thread-packages/fal FAL根据龙哥建议，应当放到主仓库
> * [x]  https://club.rt-thread.org/ask/question/433049.html
> * [x]  https://club.rt-thread.org/ask/question/431576.html
> * [x]  https://club.rt-thread.org/ask/question/430162.html
> * [x]  https://club.rt-thread.org/ask/question/430162.html
> * [ ]  libmodbus软件包编译报错问题也可以修改下，主要涉及几个头文件，论坛里已有解决方案
>   https://club.rt-thread.org/ask/article/3280.html

第三个作者修了 https://github.com/gbcwbz/canfestival-rtt/issues/9#issuecomment-1068134572

**Comment by mysterywolf** [comment]
仓库代码尽量压缩C99的使用需求，如无特殊需求使用CFLAGS或者LOCAL_CFLAGS, 而非CCFLAGS，不要讲C++和C的配置项默认混在一起。
部分编译器如（Keil-MDK），针对C++和C的配置要求不一样，混在一起配置可能会出问题。

**Comment by thewon86** [comment]
> 

为啥不是计划放弃旧版本语法，向新规范靠拢？

**Comment by mysterywolf** [comment]
> 为啥不是计划放弃旧版本语法，向新规范靠拢？

因为不同编译器对新版本语法的支持情况并不相同，因此不适合强制工程直接使用C++11或者C99，用户有需求可以通过sconscript自行定义LOCAL_CCFLAG LOCAL_CFLAG LOCAL_CXXFLAG 参见：https://github.com/mysterywolf/RTduino/blob/master/core/SConscript

**Comment by mysterywolf** [comment]
## libc
- [x] https://github.com/RT-Thread/rt-thread/issues/5708
- [x] newlib 4.1.0 新版本的支持

**Comment by mysterywolf** [comment]
版本控制规范：https://semver.org/

**Comment by mysterywolf** [comment]
> * [ ]  支持64位时间戳，以解决千年虫问题
>   
>   * [ ]  驱动支持64位时间戳，即使libc还末支持64位
>   * [ ]  libc支持64位的方案，不同的编译器现在支持的情况不同，可以考虑自己单独维护1套time相关的功能
> 
> ![2038](https://user-images.githubusercontent.com/1754905/156694274-213ce0ef-2e25-4411-a9c8-73fbf6545bc6.gif)

2038的问题 看起来不只是time32改time64这么简单 其他结构体也需要从long改成longlong https://github.com/RT-Thread/rt-thread/pull/5775#discussion_r843435505

**Comment by mysterywolf** [comment]
https://club.rt-thread.org/ask/question/435318.html
